### PR TITLE
Add Put Query Ruleset API call

### PR DESCRIPTION
**Pull request by kderusso**
Adds a PUT Query Ruleset API request. Also changes this from a system property to a feature flag. 

Example API request: 
```
PUT /_query_rules/test-ruleset
{
  "ruleset_id": "test-ruleset",
  "rules": [
    {
      "rule_id": "query-rule-id1",
      "type": "pinned",
      "criteria": [
        {
          "type": "exact",
          "metadata": "query_string",
          "value": "puggles"
        }
      ],
      "actions":
        {
          "ids": [
            "abc",
            "def"
          ]
        }
    },
    {
      "rule_id": "query-rule-id1",
      "type": "pinned",
      "criteria": [
        {
          "type": "exact",
          "metadata": "query_string",
          "value": "puggles"
        }
      ],
      "actions":
        {
          "docs": [
            { "_index": "test-index-1", "_id": "ghi"},
            { "_index": "test-index-1", "_id": "jkl"}
          ]
        }
    }
  ]
}
```

### CODE REVIEW

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearch.java
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearch.java:97]: Note: I purposely kept this external to the `_application` endpoint. (If we like this and think the classes need to move packages, I'd prefer to do that refactor/move in a single PR to limit noise)

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/200_query_ruleset_put.yml
Comment (kderusso) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/200_query_ruleset_put.yml:4]: Note: Since only the `PUT` API call is implemented and not `GET` or `DELETE`, this test is pretty lightweight. It will be extended as additional CRUD API calls are added.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/action/PutQueryRulesetAction.java
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/action/PutQueryRulesetAction.java:61]: I'm not quite sure what validation we'd want here that we don't already do in constructors. Ideas welcome?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilegeResolver.java
Comment (kderusso) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilegeResolver.java:282]: I'd especially like feedback from @ioanatia here - Is a "manage query rules" cluster privilege sufficient?

File: test/test-clusters/src/main/java/org/elasticsearch/test/cluster/FeatureFlag.java
Comment (kderusso) [inline test/test-clusters/src/main/java/org/elasticsearch/test/cluster/FeatureFlag.java:22]: I could use some input here - what test suites would benefit from having this FF enabled? Or, since the functionality is in the ent search module would it be better to remove this from `test-clusters`?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/action/PutQueryRulesetAction.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/action/PutQueryRulesetAction.java:61]: I can't think of any! Maybe we can change the TODO to a comment indicating that they are already validated in constructors

**Review by carlosdelest — COMMENTED** [review]
Some minor comments, overall 💯

File: rest-api-spec/src/main/resources/rest-api-spec/api/query_ruleset.put.json
Comment (kderusso) [reply rest-api-spec/src/main/resources/rest-api-spec/api/query_ruleset.put.json:35]: Probably not, but it's a paradigm we already use with search applications and it wasn't a big lift to support.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/action/PutQueryRulesetAction.java
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/action/PutQueryRulesetAction.java:95]: I like that idea. The only argument I can see _right now_ to keep them separate (even if it duplicates some code) is if there's any chance this API would move out of ent-search. I'm inclined to say that refactoring should be a followup PR. WDYT?

Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/action/PutQueryRulesetAction.java:95]: We can do it as a separate PR for sure

File: rest-api-spec/src/main/resources/rest-api-spec/api/query_ruleset.put.json
Comment (carlosdelest) [reply rest-api-spec/src/main/resources/rest-api-spec/api/query_ruleset.put.json:35]: I'd say YAGNI - let's not add anything we're not sure to need, to avoid testing / documenting / maintaining it. It's easy to add later if needed

Comment (kderusso) [reply rest-api-spec/src/main/resources/rest-api-spec/api/query_ruleset.put.json:35]: Most of the time I'd agree, but there is a chance of accidentally overwriting an entire ruleset, with no history. This could be a guardrail.

Comment (carlosdelest) [reply rest-api-spec/src/main/resources/rest-api-spec/api/query_ruleset.put.json:35]: I think we added that option to Search Applications for a very specific UI use case, and I don't feel this is something we need as a rule. I don't see this as a deal breaker for the PR, just a warning that we probably should not add it at least yet. It adds code and tests that we are not needing.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilegeResolver.java
Comment (ioanatia) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilegeResolver.java:282]: If it was me, I would actually just use the `manage` cluster privilege for now because this feature is behind a feature flag and it's disabled by default. So there's no immediate need to add a cluster privilege.
The `manage` cluster privilege includes all the cluster privileges except `manage_security`, so it is perfectly fine.
We can then add this as a separate cluster privilege and have a review from the Elasticsearch security team.
The `manage` cluster privilege would most likely include this new cluster privilege.

One unknown that I have would be what would happen if at some point we decide to remove or rename this `manage_query_rules` privilege - if a cluster has an API key that uses it, when it gets upgraded to a new version where this privilege does not exist, what would the behaviour of the API key be? We had a similar conversation when we thought about removing the cluster privilege we have for behavioral analytics and rely on the new workflow restrictions.

So instead of trying to answer questions like these right away, I would just not focus on it at the moment and use the `manage` cluster privilege for now.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/action/PutQueryRulesetAction.java
Comment (ioanatia) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/action/PutQueryRulesetAction.java:61]: So, I would actually try to move all the validations here if possible.
Because right now, what happens is that we validate the request payload before we validate the license.
We want the license check to be the first one that happens, not the validation of the payload.
It's a problem we have with the rest of the APIs in this module, that we did not get to fix (we have an issue for it).
But since this is a new API, maybe we can make this right from the beginning?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilegeResolver.java
Comment (kderusso) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilegeResolver.java:282]: Sounds reasonable, thank you. I'll update this 👍

**Review by carlosdelest — APPROVED** [review]

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/action/PutQueryRulesetAction.java
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/action/PutQueryRulesetAction.java]: Updated with validations here. Is that enough, @ioanatia? I don't like the idea of removing constructor validation too, but we can discuss for a followup PR

**Review by ioanatia — APPROVED** [review]

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilegeResolver.java
Comment (ioanatia) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilegeResolver.java:280]: very small nitpick that we don't need this extra line

**Pull request merged**
